packages/flutter_tools/gradle/src/main/kotlin/dependency_version_checker.gradle.kts

```diff
@@ -0,0 +1,270 @@
+            checkJavaVersion(getJavaVersion(project), project)
+            agpVersion = getAGPVersion(project)
+            if (agpVersion != null) {
+                checkAGPVersion(agpVersion, project)
+            } else {
+                project.logger.error("Warning: unable to detect project AGP version. Skipping " +
+                        "version checking. ")
+            }
+
+            kgpVersion = getKGPVersion(project)
+            if (kgpVersion != null) {
+                checkKGPVersion(kgpVersion, project)
+            } else {
+                project.logger.error("Warning: unable to detect project KGP version. Skipping " +
+                        "version checking.")
+            }
+        }
+
+        // https://docs.gradle.org/current/kotlin-dsl/gradle/org.gradle.api.invocation/-gradle/index.html#-837060600%2FFunctions%2F-1793262594
+        fun getGradleVersion(project : Project) : Version {
+            return Version.fromString(project.gradle.getGradleVersion())
+        }
+
+        // https://docs.gradle.org/current/kotlin-dsl/gradle/org.gradle.api/-java-version/index.html#-1790786897%2FFunctions%2F-1793262594
+        fun getJavaVersion(project : Project) : JavaVersion {
+            return JavaVersion.current()
+        }
+
+        // This approach is taken from AGP's own version checking plugin:
+        // https://android.googlesource.com/platform/tools/base/+/1839aa23b8dc562005e2f0f0cc8e8b4c5caa37d0/build-system/gradle-core/src/main/java/com/android/build/gradle/internal/utils/agpVersionChecker.kt#58.
+        fun getAGPVersion(project: Project): Version? {
+            var agpVersion: Version? = null
+            try {
+                agpVersion = Version.fromString(
+                    project.plugins.getPlugin("com.android.base")::class.java.classLoader.loadClass(
+                        com.android.Version::class.java.name
+                    ).fields.find { it.name == "ANDROID_GRADLE_PLUGIN_VERSION" }!!
+                        .get(null) as String
+                )
+            } catch (ignored: ClassNotFoundException) {
+                // Use deprecated Version class as it exists in older AGP (com.android.Version) does
+                // not exist in those versions.
+                agpVersion = Version.fromString(
+                    project.plugins.getPlugin("com.android.base")::class.java.classLoader.loadClass(
+                        com.android.builder.model.Version::class.java.name
+                    ).fields.find { it.name == "ANDROID_GRADLE_PLUGIN_VERSION" }!!
+                        .get(null) as String
+                )
+            }
+            return agpVersion
+        }
+
+        fun getKGPVersion(project : Project) : Version? {
+            // This property corresponds to application of the Kotlin Gradle plugin in the
+            // top-level build.gradle file.
+            if (project.hasProperty("kotlin_version")) {
```

Comment:
Consider making property strings consts with smallest scope that is needed for all references.

Reply:
Changed